File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
sys.setrecursionlimit(10**6)
from collections import deque

input = sys.stdin.readline
print = sys.stdout.write

mydeque = deque()
N, M, V = map(int, input().split())
A = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)

def DFS(v):
    print(str(v) + " ")
    visited[v] = True

    for i in A[v]:
        if not visited[i]:
            DFS(i)

def BFS(v):
    mydeque.append(v)
    visited[v] = True

    while mydeque:
        node = mydeque.popleft()
        print(str(node) + " ")

        for i in A[node]:
            if not visited[i]:
                mydeque.append(i)
                visited[i] = True

for i in range(M):
    s, e = map(int, input().split())

    A[s].append(e)
    A[e].append(s)

for i in range(N + 1):
    A[i].sort()

DFS(V)
print("\n")

visited = [False] * (N + 1)

BFS(V)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;
import java.io.*;

public class Main {
    static ArrayList<Integer>[] A;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int V = Integer.parseInt(st.nextToken());

        A = new ArrayList[N + 1];
        visited = new boolean[N + 1];

        for (int i = 1; i <= N; i++) {
            A[i] = new ArrayList<>();
        }

        for (int i = 1; i <= M; i++) {
            st = new StringTokenizer(br.readLine());

            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            A[start].add(end);
            A[end].add(start);
        }

        for (int i = 1; i <= N; i++) {
            Collections.sort(A[i]);
        }

        DFS(V);
        System.out.println();

        visited = new boolean[N + 1];

        BFS(V);
    }

    static void DFS(int v) {
        visited[v] = true;
        System.out.print(v + " ");

        for (int i : A[v]) {
            if (!visited[i]) {
                DFS(i);
            }
        }
    }

    static void BFS(int v) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(v);
        visited[v] = true;

        while (!queue.isEmpty()) {
            int now = queue.poll();
            System.out.print(now + " ");

            for (int i : A[now]) {
                if (!visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                }
            }
        }
    }
}
